File: chinese_remainder_theorem.py
# Implemented from this video
# https://youtu.be/Y5RcMWiUyyE
# Awesome video


from math import prod
from ast import literal_eval as le
import sys

def egcd(a,b):
    if a == 0 :
        return b, 0, 1
    else :
        gcd, x, y = egcd( b%a , a )
        nx = y - (b//a)*x
        ny = x
        return ( gcd, nx, ny )
        
def modinv(a,p):
    gcd, x, y = egcd(a,p)
    if gcd != 1 :
        raise Exception('Modular inverse doesn\'t exist !!')
    else:
        return x%p
        
def crt(a,n):
    '''a & n must be arrays (list) and thier
    lengths must be equal. we are trying to find 
    x ≡ a1 | mod(n1) we are trying to find
    x ≡ a2 | mod(n2) "x" such that it satisfies
    x ≡ a3 | mod(n3) the equations '''
    N = prod(n)
    sum = 0
    for i in range(len(a)):
        ai = a[i]
        ni = n[i]
        Ni = N / ni
        xi = modinv(Ni,ni)
        sum += ai*Ni*xi
    res = sum % N
    return res
    
if __name__ == '__main__' :
    a = le(sys.argv[1])
    n = le(sys.argv[2])
    print(crt(a,n))

File: c_code/chinese_remainder_theorem.c
/*
 Translated from the python in this repo
*/

#include <stdio.h>
#include <stdlib.h>
#define len(x) sizeof(x)/sizeof(*x)

typedef struct {
    int gcd ;
    int x ;
    int y ;
} ext ;

ext egcd(int a, int b);

int modinv(int a, int n);

int prod(int a[], int len);

int crt(int a[], int n[], int len);

int main(){
    /* a & n must be arrays and thier
    lengths must be equal. we are trying to find 
    x ≡ a1 | mod(n1) we are trying to find
    x ≡ a2 | mod(n2) "x" such that it satisfies
    x ≡ a3 | mod(n3) the equations */
    
    int a[] = { 3, 1, 6} ;
    int n[] = { 5, 7, 8} ;
    int ln = len(a) ;
    int result = crt(a, n, ln);
    printf("[+] x  : %d \n----------------------\n", result);
    return 0;
    }
    
ext egcd(int a,int b){
    ext val ;
    ext rev ;
    if (a == 0) {
        val.gcd = b ;
        val.x = 0 ;
        val.y = 1 ;
        return val ;
        }
    else {
        int q = b % a ;
        rev = egcd(q, a) ;
        val.gcd = rev.gcd ;
        val.x = rev.y - ((b/a)*rev.x) ;
        val.y = rev.x ;
        return val;
        }
}

int modinv(int a, int n) {
    ext res = egcd(a, n);
    if (res.gcd != 1) {
        printf("[+] Modular Inverse does not exist !") ;
        exit(1);
        }
    else {
        int result ;
        int q ;
        q = res.x ;
        result = q % n ;
        return result ;
        }
}

int prod(int arr[], int len) {
    int result = 1;
    for (int i = 0; i < len; i++){
    result = result * arr[i];
    }
    return result;
}

int crt(int a[], int n[], int len){
    int sum = 0 ;
    int N = prod(n, len) ;
    for (int i = 0; i < len; i++){
        int ai = a[i] ;
        int ni = n[i] ;
        int Ni = N/ni ;
        int xi = modinv(Ni, ni) ;
        sum += ai*Ni*xi ;
        }
    int res = sum % N ;
    return res ;
}